compiler/src/dotty/tools/dotc/quoted/PickledQuotes.scala

```diff
@@ -0,0 +1,91 @@
+package dotty.tools.dotc.quoted
+
+import dotty.tools.dotc.ast.Trees._
+import dotty.tools.dotc.ast.tpd
+import dotty.tools.dotc.config.Printers._
+import dotty.tools.dotc.core.Constants.Constant
+import dotty.tools.dotc.core.Contexts._
+import dotty.tools.dotc.core.Decorators._
+import dotty.tools.dotc.core.Flags._
+import dotty.tools.dotc.core.Symbols._
+import dotty.tools.dotc.core.tasty._
+import dotty.tools.dotc.interpreter.RawQuoted
+
+object PickledQuotes {
+  import tpd._
+
+  /** Pickle the quote into a TASTY string */
+  def pickleQuote(tree: Tree)(implicit ctx: Context): String = {
+    if (ctx.reporter.hasErrors) "<error>"
+    else {
+      val encapsulated = encapsulateQuote(tree)
+      val pickled = pickle(encapsulated)
+      TastyString.tastyToString(pickled)
+    }
+  }
+
+  /** Transform the expression into it's fully spliced Tree */
+  def quotedToTree(expr: quoted.Quoted)(implicit ctx: Context): Tree = expr match {
+    case expr: quoted.TastyQuoted => unpickleQuote(expr)
+    case expr: quoted.Liftable.PrimitiveExpr[_] => Literal(Constant(expr.value))
+    case expr: RawQuoted => expr.tree
+  }
+
+  /** Unpickle the tree contained in the TastyQuoted */
+  private def unpickleQuote(expr: quoted.TastyQuoted)(implicit ctx: Context): Tree = {
+    val tastyBytes = TastyString.stringToTasty(expr.tasty)
+    val unpickled = unpickle(tastyBytes, expr.args)
+    unpickled match { // Expects `package _root_ { val ': Any = <tree> }`
+      case PackageDef(_, (vdef: ValDef) :: Nil) => vdef.rhs
+    }
+  }
+
+  /** Encapsulate the tree in a top level `val`
+   *    `<tree>` ==> `package _root_ { val ': Any = <tree> }`
+   *
+   *  Note: Trees for types are also encapsulated this way to preserve the holes in the tree.
+   *        Encapsulating the type of the tree in a `type ' = <tree.tpe>` can potentially
```

Comment:
Did not understand why.

Reply:
Actually, I managed to encode it into a top level `type`.